packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts

```diff
@@ -0,0 +1,163 @@
+import { Counters } from '../../../common/metrics'
+import { SignerConfig } from '../../../config'
+import { PnpQuotaService } from '../../services/quota'
+import { PnpSession } from '../../session'
+import { PnpSignIO } from './io'
+import { LegacyPnpSignIO } from './io.legacy'
+
+export abstract class PnpSignAction
+  implements Action<SignMessageRequest | LegacySignMessageRequest> {
+  protected abstract readonly requestsTable: REQUESTS_TABLE
+
+  constructor(
+    readonly db: Knex,
+    readonly config: SignerConfig,
+    readonly quota: PnpQuotaService,
+    readonly keyProvider: KeyProvider,
+    readonly io: PnpSignIO | LegacyPnpSignIO
+  ) {}
+
+  public async perform(
+    session: PnpSession<SignMessageRequest | LegacySignMessageRequest>,
+    timeoutError: symbol
+  ): Promise<void> {
+    // Compute quota lookup, update, and signing within transaction
+    // so that these occur atomically and rollback on error.
+    await this.db.transaction(async (trx) => {
+      const pnpSignHandler = async () => {
+        const quotaStatus = await this.quota.getQuotaStatus(session, trx)
+
+        let isDuplicateRequest = false
+        try {
+          isDuplicateRequest = await getRequestExists(
+            this.db,
+            this.requestsTable,
+            session.request.body.account,
+            session.request.body.blindedQueryPhoneNumber,
+            session.logger,
+            trx
+          )
+        } catch (err) {
+          session.logger.error(err, 'Failed to check if request already exists in db')
+        }
+
+        if (isDuplicateRequest) {
+          Counters.duplicateRequests.inc()
+          session.logger.info(
+            'Request already exists in db. Will service request without charging quota.'
+          )
+          session.errors.push(WarningMessage.DUPLICATE_REQUEST_TO_GET_PARTIAL_SIG)
+        } else {
+          // In the case of a database connection failure, performedQueryCount will be -1
+          if (quotaStatus.performedQueryCount === -1) {
+            this.io.sendFailure(
+              ErrorMessage.DATABASE_GET_FAILURE,
+              500,
+              session.response,
+              quotaStatus
+            )
+            return
+          }
+          // In the case of a blockchain connection failure, totalQuota will be -1
+          if (quotaStatus.totalQuota === -1) {
+            if (this.config.api.phoneNumberPrivacy.shouldFailOpen) {
+              // We fail open and service requests on full-node errors to not block the user.
+              // Error messages are stored in the session and included along with the signature in the response.
+              quotaStatus.totalQuota = Number.MAX_SAFE_INTEGER
+              session.logger.warn(
+                { warning: ErrorMessage.FAILURE_TO_GET_TOTAL_QUOTA },
+                ErrorMessage.FAILING_OPEN
+              )
+              Counters.requestsFailingOpen.inc()
```

Comment:
Is it possible to add a check to this code and have the service start failing closed if it reaches a threshold? One thing to note is that if an attacker were trying to compromise the system with this route (e.g. they are DoSing the full node and requesting directly to the signer) the combiner will not see anything happening.

Reply:
I think this is unfortunately non-trivial, though we do have [a ticket](https://github.com/celo-org/celo-monorepo/issues/9835) tracking this and want to more fully explore options around the fail open logic post-merge. @alecps any additional thoughts?

Reply:
- Decision after talking with victor is to have this default to closed